Adding the dry-bulb air temperature and tv flag in the observation diagnostic file of conventional Q obs (for 3D RTMA run only)

GangZhao-NOAA opened the pull request:
**DUE DATE for merger of this PR into `develop` is 3/1/2024 (six weeks after PR creation).**


**Description**

<!-- Please include relevant motivation and context. -->
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- List any dependencies that are required for this change. -->

- Motivation:
In the Auto-QC utility of 3D-RTMA, the dry-bulb air temperature (hereafter as T_dry) which is accompanied with moisture observation is required in the procedure of the quality control for conventional moisture observations (Q obs). Since this auto-qc of 3DRTMA retrieves many information from the observation diagnostic files (output of GSI run), it should be convenient for the sake of auto-qc of Q obs, if the required T_dry could be output into the observation diagnostics file of Q obs in the GSI run.

- Modifications to the code:
**_read_prepbufr.f90_**:
for qob and rtma run, add lines to use tpc to identify the sensible temp and virtual temp, and mark with tvflag, then get tdry with the temp obs in different ways w.r.t the associated tvflag;
**_setupq.f90_**:
add line to use l_rtma3d from module rapidrefresh_cldsurf_mod
variable nreal (the length of obs diag information record for each obs) needs to be increased by 2 (for tdry and tvflag, when running GSI for 3DRTMA only)
in subroutine contents_binary_diag_, add line to put tdry into the array rdiagbuff  (for binary format obsdaig file)
in subroutine contents_netcdf_diag_, add line to put tdry into metadata (for netcdf format obsdiag file)

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->
This PR is to address the issue #666 : Adding the (calculated) dry-bulb temperature in the observation diagnostic file for conventional Q obs (only for 3D RTMA)
Fixes #666

**Type of change**

Please delete options that are not relevant.

- [* ] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->
  
**Checklist**

- [*] My code follows the style guidelines of this project
- [*] I have performed a self-review of my own code
- [*] I have commented my code, particularly in hard-to-understand areas
- [*] New and existing tests pass with my changes
- [*] Any dependent changes have been merged and published
- [*] Regression Test of GSI

      1. The modified GSI code passed the regression test of GSI on WCOSS2 and Hera. Here is a brief report of the ctest run on WCOSS2:
```
[gang.zhao@clogin09:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[gang.zhao@clogin09:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.29 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  484.26 sec
3/7 Test #7: global_enkf ......................   Passed  607.50 sec
4/7 Test #2: rtma .............................   Passed  966.51 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1088.44 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1208.36 sec
7/7 Test #1: global_4denvar ...................   Passed  1322.12 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1322.12 sec
```
      2. results from ctest running on Hera:
 ```
 [Gang.Zhao@hfe10:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[Gang.Zhao@hfe10:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............***Failed  551.60 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  608.31 sec
3/7 Test #7: global_enkf ......................   Passed  1035.43 sec
4/7 Test #2: rtma .............................   Passed  1636.12 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1646.84 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1704.16 sec
7/7 Test #1: global_4denvar ...................   Passed  1795.47 sec

86% tests passed, 1 tests failed out of 7

Total Test time (real) = 1795.49 sec

The following tests FAILED:
          3 - rrfs_3denvar_glbens (Failed)
Errors while running CTest
[Gang.Zhao@hfe10:build] (feature/tdry_in_qob_diag_rtma)$ ctest -R rrfs_3denvar_glbens
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 3: rrfs_3denvar_glbens
1/1 Test #3: rrfs_3denvar_glbens ..............   Passed  1099.08 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 1099.11 sec
```

- [*] The modified GSI code has been tested with an analysis case on Hera (for RRFS-based 3DRTMA case 2023-12-16_12Z)
``` In the observation diagnostic file for q-obs generated by GSI, T_dry and tvflag were added for each record of Q obs in binary format diagnostic file; and in netcdf-4 format diagnostic file, new metadata variable "Observation_Tdry" and "Setup_QC_Mark" (used for tvflag) were added in the metadata group. The other variables in obsdiag file are still as same as in the obsdiag file from original GSI code. It indicates that the modification does not affect the other variables in the obsdiag files. And in the test run with l_rtma3d=.false. (this is the GSI run not for 3D RTMA), then in the obsdiag file for Q obs, there is no Tdry and tvflag neither in binary format nor in netcdf-4 format, this is also as expected that this modifications should have no influence for other GSI applications. And the most impratant thing is, the analysis, minimisation information, obs-fitting stats from the control run (using original code without adding tdry) and testing run with new code adding tdry are identical. So the modified code have no influences to the analysis. ```

GangZhao-NOAA committed:
Add dry-bulb temperature and tv flag in q obsdiag file when running
RTMA.

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop'(#7f37de3) into feature/tdry_in_qob_diag_rtma

ShunLiu-NOAA commented:
@MatthewMorris-NOAA @xyzemc and @guoqing-noaa   Could you review this PR? Thanks.

GangZhao-NOAA committed:
write out tdry and tvflag in binary format q-obsdiag for 2DRTMA as well.

GangZhao-NOAA commented:
Hi, @MatthewMorris-NOAA @ManuelPondeca-NOAA @xyzemc @guoqing-noaa @ShunLiu-NOAA,
Following Manuel's suggestion, I updated the code (just very minor changes) to add the tdry & tv flag to the q-obsdiag for 2DRTMA run. The updated GSI code also passed the ctest on Hera and WCOSS2 both.
Here is the results from ctest on WCOSS2:
```
gang.zhao@clogin05:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_Dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.42 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  484.29 sec
3/7 Test #7: global_enkf ......................   Passed  608.24 sec
4/7 Test #2: rtma .............................   Passed  968.85 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1089.00 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1208.67 sec
7/7 Test #1: global_4denvar ...................   Passed  1321.89 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1321.89 sec
```

Now you may start to review the updated code.
Thank you!

xyzemc commented:
> @MatthewMorris-NOAA @xyzemc and @guoqing-noaa Could you review this PR? Thanks.

sure

AnnetteGibbs-NOAA reviewed (approved):
The code changes look good to me.

GangZhao-NOAA commented:
Due to the latest commit #[8ed034f](https://github.com/NOAA-EMC/GSI/commit/8ed034fe6f2a1aac5ce145b4a1b441e3aa37d8ca) (103 files were changed), which cause some conflicts with my modifications in read_prepbufr.f90. I need solve these conflicts and re-run the tests. Close this PR #688 for now.
Thank you!

GangZhao-NOAA committed:
Merge commit #8ed034fe6f2a1aac5ce145b4a1b441e3aa37d8ca
remote-tracking branch 'upstream/develop' into feature/tdry_in_qob_diag_rtma

AnnetteGibbs-NOAA reviewed (approved):
Looks good Gang!

GangZhao-NOAA commented:
Hi @AnnetteGibbs-NOAA @MatthewMorris-NOAA,
I fixed the code conflicts caused by latest GSI updates (PR#670 commit#[8ed034f](https://github.com/NOAA-EMC/GSI/commit/8ed034fe6f2a1aac5ce145b4a1b441e3aa37d8ca)).
Could you please make another round of code review for this PR#688? I will put the ctest results very soon.

Thank you!

GangZhao-NOAA commented:
Thank you, @AnnetteGibbs-NOAA !

GangZhao-NOAA commented:
with my latest updated code (for this PR#688), new ctest (regression tests) running Hera:
~~~
[Gang.Zhao@hfe02:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[Gang.Zhao@hfe02:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  488.65 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  495.18 sec
3/7 Test #7: global_enkf ......................   Passed  942.42 sec
4/7 Test #2: rtma .............................   Passed  973.65 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1229.37 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1354.77 sec
7/7 Test #1: global_4denvar ...................   Passed  1674.00 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1674.03 sec

Regression Results:  /scratch1/NCEPDEV/da/Gang.Zhao/noscrub/regression_61a7a8f_tdry_on_8ed034f_hera

~~~

MatthewMorris-NOAA reviewed (approved):
Looks good!

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit#a898668)
into feature/tdry_in_qob_diag_rtma

GangZhao-NOAA commented:
This morining EMC GSI was updated with the latest [commit #a898668](https://github.com/GangZhao-NOAA/GSI/commit/a8986684a815a9f2602a11e962f8443b78169012). So I updated my fork of GSI and all the branches, including working branch "feature/tdry_in_qob_diag_rtma" to the current branch head of develop. As for now, my working branch for this PR #688 is 5 commits ahead of develop branch of EMC GSI.

Since the changes of commit#a898668 were related to the build files of GSI, there is No any changes to source files. So I believe after I updated to current head of develop branch, my code should be good enough, there is no need for further peer review and ctest. But if the ctests are still necessary, please let me know.

Thank you!

GangZhao-NOAA commented:
ctest (7 regression tests) were done today on wcoss2 and hera with the latest updates (after merging latest EMC GSI commit#a898668 into working branch "feature/tdry_in_qob_diag_rtma" (commit #b11bedd) of my personal fork). All the regression tests were passed successfully.
Here are the results:
on wcoss2  (dogwood):
~~~
[gang.zhao@dlogin01:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[gang.zhao@dlogin01:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.52 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  484.85 sec
3/7 Test #7: global_enkf ......................   Passed  607.84 sec
4/7 Test #2: rtma .............................   Passed  966.52 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1090.70 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1208.75 sec
7/7 Test #1: global_4denvar ...................   Passed  1384.29 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1384.29 sec

GangZhao-NOAA:feature/tdry_in_qob_diag_rtma  commit #b11bedd running on wcoss2/dogwood
~~~

on hera:
~~~
[Gang.Zhao@hfe12:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[Gang.Zhao@hfe12:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  489.14 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  967.69 sec
3/7 Test #6: hafs_3denvar_hybens ..............   Passed  1223.83 sec
4/7 Test #2: rtma .............................   Passed  1391.67 sec
5/7 Test #7: global_enkf ......................   Passed  1409.88 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1526.33 sec
7/7 Test #1: global_4denvar ...................   Passed  1672.14 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1672.16 sec

GangZhao-NOAA:feature/tdry_in_qob_diag_rtma  commit #b11bedd running on hera
~~~

hu5970 commented:
@GangZhao-NOAA Could you sync this PR with develop and do regression tests. I think this PR is ready for merging. Thanks,
Ming

GangZhao-NOAA committed:
Merge remote-tracking branch 'upstream/develop' (commit #bae0342)
into feature/tdry_in_qob_diag_rtma

GangZhao-NOAA commented:
@hu5970 Hi, Ming,

Following your suggestion, I merged the latest EMC GSI develop branch (commit #bae0342f) into my working branch "feature/tdry_in_qob_diag_rtma" (the current commit #26a5401). So now it is updated to the current head of branch develop. Then I ran the ctest (regression tests) on Hera and WCOSS2 (Dogwood). All the 7 tasks passed successfully on both machines. 
Thank you!

Here are the test results:

on Hera:
~~~
[Gang.Zhao@hfe05:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[Gang.Zhao@hfe05:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Gang.Zhao/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  485.46 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  487.99 sec
3/7 Test #7: global_enkf ......................   Passed  959.74 sec
4/7 Test #2: rtma .............................   Passed  1046.67 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1127.61 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1361.19 sec
7/7 Test #1: global_4denvar ...................   Passed  1625.24 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1625.26 sec
~~~

On WCOSS2 (Dogwood):
~~~
[gang.zhao@dlogin05:build] (feature/tdry_in_qob_diag_rtma)$ ctest -N
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
  Test #1: global_4denvar
  Test #2: rtma
  Test #3: rrfs_3denvar_glbens
  Test #4: netcdf_fv3_regional
  Test #5: hafs_4denvar_glbens
  Test #6: hafs_3denvar_hybens
  Test #7: global_enkf

Total Tests: 7
[gang.zhao@dlogin05:build] (feature/tdry_in_qob_diag_rtma)$ ctest -j 7
Test project /lfs/h2/emc/da/save/gang.zhao/WorkDir/ProdGSI_dev/gsi_dev/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.96 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  485.16 sec
3/7 Test #7: global_enkf ......................   Passed  611.18 sec
4/7 Test #2: rtma .............................   Passed  967.83 sec
5/7 Test #5: hafs_4denvar_glbens ..............   Passed  1209.90 sec
6/7 Test #6: hafs_3denvar_hybens ..............   Passed  1209.92 sec
7/7 Test #1: global_4denvar ...................   Passed  1323.27 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1323.28 sec
~~~

hu5970 reviewed (approved):
The code looks good.
Two reviewers approved this PR. All regression test cases passed.

RussTreadon-NOAA commented on src/gsi/read_prepbufr.f90:
```diff
@@ -149,8 +149,10 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
 !   2020-09-05  CAPS(C. Tong) - add flag for new vadwind obs to assimilate around the analysis time only
 !   2023-03-23  draper  - add code for processing T2m and q2m for global system
-!   2023-07-30  Zhao    - added code to extract obs of significant wave height (howvob) from bufr record 
+!   2023-07-30  zhao    - added code to extract obs of significant wave height (howvob) from bufr record 
```
General comment:  Since our code is version controlled we do not need to add entries to the program history log.  We can use git commands to see what, when, and who changed code.

GangZhao-NOAA replied:
@RussTreadon-NOAA
I was wondering why there are no more these entries for the recent GSI updates. Now I see.
Thank you!
-Gang

RussTreadon-NOAA commented on src/gsi/read_prepbufr.f90:
```diff
@@ -1713,6 +1724,25 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
+                 tobs4q(1,k)=obsdat(3,k)                      ! temp obs read in prepbufr
+                 tqm4q(1,k)=tqm(k)
+                 do j=1,20
+                    if (tpc(k,j)==vtcd) tvflg4q(1,k)=zero     ! reset flag if virtual
```
General comment:  We should minimize the use of `==` when comparing floating point values.  This is done elsewhere in `read_prepbufr.f90` so no change needed.  Just a general comment.

GangZhao-NOAA replied:
I will avoid comparing two real numbers with "==" in the future code work.
Thank you.
-Gang

RussTreadon-NOAA commented:
Two minor comments @GangZhao-NOAA . You don't need to change code in this PR.  My comments are for future PRs.

Question:  Have you or others confirmed that the additional information written to the binary and netcdf diagnostic files is does not cause any problems for downstream codes, specifically observation monitoring codes.   Related question:  have we confirmed that the additional information written to the diagnostic files is correct?  

I ask these questions since none of our ctests check or use the diagnostic file created by the updated `gsi.x`.   Getting the 7 ctests to pass is good but it doesn't tell us if the changes in this PR work as intended.

GangZhao-NOAA commented:
> Two minor comments @GangZhao-NOAA . You don't need to change code in this PR. My comments are for future PRs.
> 
> Question: Have you or others confirmed that the additional information written to the binary and netcdf diagnostic files is does not cause any problems for downstream codes, specifically observation monitoring codes. Related question: have we confirmed that the additional information written to the diagnostic files is correct?
> 
> I ask these questions since none of our ctests check or use the diagnostic file created by the updated `gsi.x`. Getting the 7 ctests to pass is good but it doesn't tell us if the changes in this PR work as intended.

@RussTreadon-NOAA , Hi, Russ,
The modifications to diagnostic files in GSI are control with logical variables "l_rtma3d" and "twodvar_regional". Only when running for 3drtma or 2drtma, the diagnostic file for q would be modified (with tdry being added). So the main DA applications (GDAS, RDAS, chem-DA, etc.) should be fine.
For 2DRTMA run, we do need to modify the downstream code which reads the diagnostic files. I've already made one and my colleague is testing it. It looks good.

Thank you!

-Gang

RussTreadon-NOAA commented:
Good to get confirmation that the 2DRTMA diagnostic file is correct.  Hopefully the same remark applies to a 3DRTMA run using the changes in this PR.

GangZhao-NOAA commented:
> Good to get confirmation that the 2DRTMA diagnostic file is correct. Hopefully the same remark applies to a 3DRTMA run using the changes in this PR.
@RussTreadon-NOAA 
Hi Russ,
Yes, it also works well for 3DRTMA run. I didn't mention 3DRTMA in the previous comment because 3DRTMA is not in operational suites yet. Actually 3DRTMA, similar as main DA systems (GDAS and RDAS at EMC), is using the netcdf format obs-diag file, so it is easy to handle with the new variables in obs-diag file. As for 2DURMA, it is still using binary-format diagnostic file, user has to use more caution and search in the corresponding setup code of GSI for the exact location of the new variable in binary obs-diag file.
Thank you! 
-Gang
